platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/configuration/problems/PropertyProblem.kt

```diff
@@ -107,60 +110,101 @@ data class StructuredMessage(val fragments: List<Fragment>) {
 }
 
 
+/**
+ * Subtypes are expected to support [PropertyTrace.equals] and [PropertyTrace.hashCode].
+ *
+ * Subclasses also must provide custom `toString()` implementations,
+ * which should invoke [PropertyTrace.asString].
+ */
 sealed class PropertyTrace {
 
-    object Unknown : PropertyTrace()
+    object Unknown : PropertyTrace() {
+        override fun toString(): String = asString()
+        override fun equals(other: Any?): Boolean = other === this
+        override fun hashCode(): Int = 0
+    }
 
-    object Gradle : PropertyTrace()
+    object Gradle : PropertyTrace() {
+        override fun toString(): String = asString()
+        override fun equals(other: Any?): Boolean = other === this
+        override fun hashCode(): Int = 0
```

Comment:
💭 It feels cleaner to use a different hash code here.
```suggestion
        override fun hashCode(): Int = 1
```

Reply:
@bamboo What about 

``` 
override fun hashCode(): Int = Gradle::class.hashCode()
```

?

Reply:
❓ By the way, aren't the default implementations of `equals` and `hashCode` enough? Why do they have to be abstract in the supertype?

Reply:
They are enough, indeed, at least of the `object`-based implementations. I wanted to signal that subclasses need to override them (as we do for `toString()`), even if all current subclasses did already get proper implementations (via `object` or `data class`), but if any of them needed to stop being a `data class` or `object`, we could forget to implement `equals` and `hashCode`.

Reply:
> I wanted to signal that subclasses need to override them
Ok 👍